rslib/src/collection/backup.rs

```diff
@@ -0,0 +1,477 @@
+    zip.start_file("collection.anki21b", options)?;
+    zstd_copy(&mut col_data, &mut zip)?;
+    zip.start_file("media", options)?;
+    zip.write_all(b"{}")?;
+    zip.finish()?;
+
+    Ok(())
+}
+
+/// Copy contents of reader into writer, compressing as we copy.
+fn zstd_copy<R: Read, W: Write>(reader: &mut R, writer: &mut W) -> Result<()> {
+    let mut encoder = Encoder::new(writer, 0)?;
+    encoder.multithread(num_cpus::get() as u32)?;
+    io::copy(reader, &mut encoder)?;
+    encoder.finish()?;
+    Ok(())
+}
+
+fn thin_backups<P: AsRef<Path>>(backup_folder: P, limits: Backups) -> Result<()> {
+    let backups =
+        read_dir(backup_folder)?.filter_map(|entry| entry.ok().and_then(Backup::from_entry));
+    let obsolete_backups = BackupThinner::new(Local::today(), limits).thin(backups);
+    for backup in obsolete_backups {
+        let _ = remove_file(backup.path);
+    }
+
+    Ok(())
+}
+
+fn out_path<S: AsRef<OsStr>>(backup_folder: S) -> PathBuf {
+    Path::new(&backup_folder).join(&format!("{}", Local::now().format(BACKUP_FORMAT_STRING)))
+}
+
+fn datetime_from_file_name(file_name: &str) -> Option<DateTime<Local>> {
+    NaiveDateTime::parse_from_str(file_name, BACKUP_FORMAT_STRING)
+        .ok()
+        .and_then(|datetime| Local.from_local_datetime(&datetime).latest())
+}
+
+#[derive(Debug, PartialEq, Clone)]
+struct Backup {
+    path: PathBuf,
+    datetime: DateTime<Local>,
+}
+
+impl Backup {
+    /// Serial day number
+    fn day(&self) -> i32 {
+        self.datetime.num_days_from_ce()
+    }
+
+    /// Serial week number, starting on Monday
+    fn week(&self) -> i32 {
+        // Day 1 (01/01/01) was a Monday, meaning week rolled over on Sunday (when day % 7 == 0).
+        // We subtract 1 to shift the rollover to Monday.
+        (self.day() - 1) / 7
+    }
+
+    /// Serial month number
+    fn month(&self) -> u32 {
+        self.datetime.year() as u32 * 12 + self.datetime.month()
+    }
+}
+
+impl Backup {
+    fn from_entry(entry: DirEntry) -> Option<Self> {
+        entry
+            .file_name()
+            .to_str()
+            .and_then(datetime_from_file_name)
+            .map(|datetime| Self {
+                path: entry.path(),
+                datetime,
+            })
+    }
+}
+
+#[derive(Debug)]
+struct BackupThinner {
+    yesterday: i32,
+    last_kept_day: i32,
+    last_kept_week: i32,
+    last_kept_month: u32,
+    limits: Backups,
+    obsolete: Vec<Backup>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum BackupStage {
+    Daily,
+    Weekly,
+    Monthly,
+}
+
+impl BackupThinner {
+    fn new(today: Date<Local>, limits: Backups) -> Self {
+        Self {
+            yesterday: today.num_days_from_ce() - 1,
+            last_kept_day: i32::MAX,
+            last_kept_week: i32::MAX,
+            last_kept_month: u32::MAX,
+            limits,
+            obsolete: Vec::new(),
+        }
+    }
+
+    /// Returns a [Vec] with the obsolete [Backup]s.
```

Comment:
We're kind of restating the type signature here. How about:

```suggestion
    /// Returns obsolete backups that can be deleted.
```

We should probably tweak the method name too, since it no longer does what it original did. Maybe "obsolete_backups()"? Then we no longer need the docstring at all.